### ISSUE TITLE
[ENH]: Make "No artists with labels found to put in legend" a warning

### ISSUE THREAD

**Issue by Jacob-Stevens-Haas** [matplotlib/matplotlib#27145]
### Problem

It's hard to debug and find the specific call to `ax.legend()` when I receive this issue if I can't raise Warnings to Exceptions.  I sometimes nest plotting functions in order to draw the same plot different ways or arrange plots in a composite figure, depending upon the presentation.

Full message
> No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.

### Proposed solution

Make this a true `UserWarning`.

Additional pro: In the future, it would be easier to find and work on this warning if it is, in fact, a true `Warning`.

**Comment by jklymak** [comment]
This is related to #23422 though this one is a little more ambiguous if it should actually be a warning or not as the user _can_ suppress the warning by either adding a label or removing the legend call.

**Comment by Jacob-Stevens-Haas** [comment]
Thanks for the link :)  My issue is definitely an example of @mwaskom's point that "you can convert a warning to an error and then drop into a debugger to figure out what's happening"


--------------------------------------
Also, I'm not sure I understand how its ambiguous?  Your statement:

> the user can suppress the warning by either adding a label or removing the legend call.

sounds like the standard, "the issue is avoidable and the client application should be modified to eliminate the warning."  I can't think of a situation where a user would legitimately intend to call `.legend()` without labels.  Matplotlib won't draw a legend when it triggers the warning.

**Comment by jklymak** [comment]
I agree, but I don't remember the original rationale for a logging call versus warning, and the rule itself is a bit ambiguous.

**Comment by timhoffm** [comment]
Thanks for the archeology 🚀! I conclude that the current state has evolved organically and there has probably not been a conscious decision between logging and warnings for this specific case.

Either way,

> I can't think of a situation where a user would legitimately intend to call .legend() without labels.

plus

> the user can suppress the warning by either adding a label or removing the legend call.

is a clear indication for me that this should be a `UserWarning`.

**Comment by jklymak** [comment]
@Jacob-Stevens-Haas Did you have bandwidth to put in a PR to change this?

**Comment by Jacob-Stevens-Haas** [comment]
Am I supposed to add a file to [matplotlib/doc/users/next_whats_new](https://github.com/matplotlib/matplotlib/tree/main/doc/users/next_whats_new) for this kind of change?
